joomla/joomla-cms
# [5.1] SEF: Implementing trailing slash behavior

### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is the behavior of trailing slashes in URLs. Right now, Joomla has largely been following a no-trailing-slashes policy, however we never enforced this. At the same time, some SEO people wanted to have trailing slashes on all URLs. This PR introduces this feature.

This introduces 2 new settings in the SEF system plugin. The first setting lets you decide if you want to have a trailing slash, no trailing slash or no special behavior (which is the b/c setting). Depending on the setting, it either adds or removes all trailing slashes.

The second setting is available when one of the former options is set and redirects a GET request with missing or trailing slash to the "correct" URL with a 301 depending on the other option.

This PR depends on #42692.

I'd like to thank [djumla GmbH](https://www.djumla.de) for sponsoring this feature.


### Testing Instructions
1. Apply the PR
2. Check that no URLs changed
3. Edit libraries/src/Router/SiteRouter.php and change line 466 to `$uri->setPath($uri->getPath() . '/' . $tmp . '/');`
4. See that all URLs now contain a trailing slash.
5. Go to the SEF system plugin in the backend and set the setting to remove the trailing slash.
6. See that all URLs now don't contain a trailing slash.
7. Revert the change from step 3.
8. Change the setting in the SEF plugin to always add a trailing slash.
9. See that all URLs now contain a traling slash again.
10. Set the second option to enforce the behavior with a 301 redirect.
11. Go to the frontend again and either add or remove the trailing slash, depending on your setting, in your browser bar.
12. See that Joomla redirected you to the same page with the correct URL again.



### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
